techmatters/hrm
# CSAM: Attach Self-Generated Report to Contact

Primary reviewer: @stephenhand 

## Description
This PR:
- Adds a migration to update `CSAMReports` table:
  - Adds column `reportType` (text) to specify the type of the report (`self-generated` or `counsellor-generated`).
  - Adds column `aknowledged` (boolean) to indicate if the report was aknowledged by the external organization. This is only relevant for `self-generated` reports, since we want to confirm that the record in the DB also has a match in the outer service. All `counsellor-generated` records are aknowledged at creation time, since they are saved in the DB **_after_** they are reported to IWF.
  - Update all existing records.
- Add endpoints and logic to support two new operations: `aknowledgeCsamReport` and `deleteCsamReport` (they do what the name says, nothing more :P).
- Extend support for `createCSAMReport` to accept `reportType` = `self-generated`. In this case we generate a UUID that will replace the `csamReportId` value (that is populated with the one generated by IWF in the case of `cousellor-generated` reports).

Q: Is desired that we filter out the reports if they are not acknowledged?

### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1407)

### Verification steps
~A PR on the Flex side will be up tomorrow, which will make the testing way easier :)~
It's a draft, but you could use https://github.com/techmatters/hrm/pull/268 to see if this is working :) (see instructions there).
- Start containerized DB, run migrations, compile code and start local server.
- Confirm that the `POST` operation on `/csamReports` works as expected for the existing csam reports logic.
- Confirm that the `POST` operation on `/csamReports` works as expected when the payload contains `reportType` = `self-generated`.
- Confirm that the `PATCH` operation on `/csamReports/:reportId/aknowled` sets `aknowledged` = `true` (for `counsellor-generated` reports is a no-op).
- Confirm that the `DELETE` operation on `/csamReports/:reportId` deletes the contact if exists.
